---
title: "Manuscript Figures for methylCC"
author: "Stephanie Hicks"
date: "`r Sys.Date()`"
output: 
    html_document:
        toc: true
        toc_float: true
        highlight: tango
        number_sections: true
        code_folding: hide
---

Load libraries and helper functions

```{r loadLibraries}
library(tidyr)
library(dplyr)
library(rafalib)
library(minfi)
library(bsseq)
library(quadprog)
library(methylCC)
library(GenomicRanges)
library(FlowSorted.Blood.450k)
library(ggplot2)
library(cowplot)
library(Gviz)

data(offMethRegions)
data(onMethRegions)

# modified function from minfi::estimateCellCounts() to return 600 Houseman CpGs
estimateCellCounts.modified <- function (rgSet, compositeCellType = "Blood", 
                                         processMethod = "auto", probeSelect = "auto",
                                cellTypes = c("CD8T","CD4T", "NK","Bcell","Mono","Gran"),
                                referencePlatform = c("IlluminaHumanMethylation450k",
                                                      "IlluminaHumanMethylationEPIC",
                                                      "IlluminaHumanMethylation27k"),
                                returnAll = FALSE, meanPlot = FALSE, verbose=TRUE, ...) {
    
    minfi:::.isRGOrStop(rgSet)
    rgSet <- as(rgSet, "RGChannelSet")
    referencePlatform <- match.arg(referencePlatform)
    rgPlatform <- sub("IlluminaHumanMethylation", "", annotation(rgSet)[which(names(annotation(rgSet))=="array")])
    platform <- sub("IlluminaHumanMethylation", "", referencePlatform)
    if((compositeCellType == "CordBlood") && (!"nRBC" %in% cellTypes))
        message("[estimateCellCounts] Consider including 'nRBC' in argument 'cellTypes' for cord blood estimation.\n")   
    referencePkg <- sprintf("FlowSorted.%s.%s", compositeCellType, platform)
    subverbose <- max(as.integer(verbose) - 1L, 0L)
    if(!require(referencePkg, character.only = TRUE))
        stop(sprintf("Could not find reference data package for compositeCellType '%s' and referencePlatform '%s' (inferred package name is '%s')",
                     compositeCellType, platform, referencePkg))
    data(list = referencePkg) 
    referenceRGset <- get(referencePkg)
    if(rgPlatform != platform) {
        rgSet <- convertArray(rgSet, outType = referencePlatform, verbose = subverbose)
    }
    if(! "CellType" %in% names(colData(referenceRGset)))
        stop(sprintf("the reference sorted dataset (in this case '%s') needs to have a phenoData column called 'CellType'"),
             names(referencePkg))
    if(sum(colnames(rgSet) %in% colnames(referenceRGset)) > 0)
        stop("the sample/column names in the user set must not be in the reference data ")
    if(!all(cellTypes %in% referenceRGset$CellType))
        stop(sprintf("all elements of argument 'cellTypes' needs to be part of the reference phenoData columns 'CellType' (containg the following elements: '%s')",
                     paste(unique(referenceRGset$cellType), collapse = "', '")))
    if(length(unique(cellTypes)) < 2)
        stop("At least 2 cell types must be provided.")
    if ((processMethod == "auto") && (compositeCellType %in% c("Blood", "DLPFC")))
        processMethod <- "preprocessQuantile"
    if ((processMethod == "auto") && (!compositeCellType %in% c("Blood", "DLPFC")))
        processMethod <- "preprocessNoob"
    processMethod <- get(processMethod)
    if ((probeSelect == "auto") && (compositeCellType == "CordBlood")){
        probeSelect <- "any"} 
    if ((probeSelect == "auto") && (compositeCellType != "CordBlood")){
        probeSelect <- "both"}
    
    if(verbose) message("[estimateCellCounts] Combining user data with reference (flow sorted) data.\n")
    newpd <- DataFrame(sampleNames = c(colnames(rgSet), colnames(referenceRGset)),
                       studyIndex = rep(c("user", "reference"),
                                        times = c(ncol(rgSet), ncol(referenceRGset))),
                       stringsAsFactors = FALSE)
    referencePd <- colData(referenceRGset)
    combinedRGset <- combineArrays(rgSet, referenceRGset, outType = "IlluminaHumanMethylation450k")
    colData(combinedRGset) <- newpd
    colnames(combinedRGset) <- newpd$sampleNames
    rm(referenceRGset)
    
    if(verbose) message("[estimateCellCounts] Processing user and reference data together.\n")
    if (compositeCellType == "CordBlood"){
        ## Here Shan wants to discard probes that they have decided shouldn't be used, for example multi-mapping probes
        ## This is done by only using probes with names in the comptable.
        ## This is kind of ugly, and dataset dependent.
        combinedMset <- processMethod(combinedRGset, verbose=subverbose)
        compTable <- get(paste0(referencePkg, ".compTable"))
        combinedMset <- combinedMset[which(rownames(combinedMset) %in% rownames(compTable)),]
    } else {
        combinedMset <- processMethod(combinedRGset) 
    }
    rm(combinedRGset)
    
    ## Extracts normalized reference data 
    referenceMset <- combinedMset[, combinedMset$studyIndex == "reference"]
    colData(referenceMset) <- as(referencePd, "DataFrame")
    mSet <- combinedMset[, combinedMset$studyIndex == "user"]
    colData(mSet) <- as(colData(rgSet), "DataFrame")
    rm(combinedMset)
    
    if(verbose) message("[estimateCellCounts] Picking probes for composition estimation.\n")
    compData <- minfi:::pickCompProbes(referenceMset, cellTypes = cellTypes, compositeCellType = compositeCellType, probeSelect = probeSelect)
    coefs <- compData$coefEsts
    rm(referenceMset)
    
    if(verbose) message("[estimateCellCounts] Estimating composition.\n")
    counts <- minfi:::projectCellType(getBeta(mSet)[rownames(coefs), ], coefs)
    rownames(counts) <- colnames(rgSet)
    
    if (meanPlot) {
        smeans <- compData$sampleMeans
        smeans <- smeans[order(names(smeans))]
        sampleMeans <- c(colMeans(getBeta(mSet)[rownames(coefs), ]), smeans)
        
        sampleColors <- c(rep(1, ncol(mSet)), 1 + as.numeric(factor(names(smeans))))
        plot(sampleMeans, pch = 21, bg = sampleColors)
        legend("bottomleft", c("blood", levels(factor(names(smeans)))),
               col = 1:7, pch = 15)
    }
    if(returnAll) {
        list(counts = counts, coefs = coefs, compTable = compData$compTable,
             normalizedData = mSet)
    } else {
        counts
    }
}

gg_color_hue <- function(n) {
  hues = seq(15, 375, length=n+1)
  hcl(h=hues, l=65, c=100)[1:n]
}

dataPathProject <- "/users/shicks1/projects/methylCC"
dataPathCarmona <- "/users/shicks1/projects/methylCC/dataFromCarmona-RRBS-450k"
dataPathLiu <- "/users/shicks1/projects/methylCC/dataFromLiu2013"

```


# Figure 1

### Load Carmona et al. (2017) data

```{r loadData-Carmona}
# load 450K data
load(file.path(dataPathCarmona, "mset450krawall.RData"))
mset450krawall <- updateObject(mset450krawall)
RGset_target450k <- mset450krawall
rm(mset450krawall)

# This is needed to run the estimateCellCounts.modified() function
colData(RGset_target450k)$Sentrix_Position <- as.character(colData(RGset_target450k)$Sentrix_Position)

# add ID to pData()
library(stringr)
newnames <- sapply(pData(RGset_target450k)$sampleID, function(x){
  str_split(x, pattern = "_")[[1]][1] })
pData(RGset_target450k)$ID <- newnames
RGset_target450k <- RGset_target450k[, !(pData(RGset_target450k)$ID %in% "D07repC")]

# create target450k datasets
Mset_target450k <- preprocessIllumina(RGset_target450k)
Mset_target450k <- mapToGenome(Mset_target450k, mergeManifest = FALSE)

pd_target450k = as.data.frame(pData(Mset_target450k))
p_target450k = getBeta(Mset_target450k , type = "Illumina")
gr_target450k <- granges(Mset_target450k)

mcols(gr_target450k) <- p_target450k

# load RRBS data
BS <- readRDS(file.path(dataPathCarmona, "bsseqkrawall.RDS"))

cvg_targetbs <- as.data.frame(getBSseq(BS, type = "Cov"))
M_targetbs <- as.data.frame(getBSseq(BS, type = "M"))
p_targetbs <- as.data.frame(getMeth(BS, type="raw")) # meth / cvg
p_targetbs <- as.data.frame(p_targetbs)
pd_targetbs <- pData(BS)
gr_targetbs <- granges(BS)
mcols(gr_targetbs) <- p_targetbs

```

### Process Carmona et al. (2017) data

```{r Houseman-Carmona}
# run estimateCellCounts.modified() on 450k samples (returns 600 Houseman CpGs)
results.450k.Houseman <- estimateCellCounts.modified(RGset_target450k, returnAll = TRUE)
counts.450k.Houseman <- results.450k.Houseman$counts / rowSums(results.450k.Houseman$counts)
rownames(counts.450k.Houseman) <- pData(RGset_target450k)$sampleID

# write.csv(counts.450K.Houseman, file = file.path(dataPathCarmona, "cellcounts/450k_minfi-estimateCellCounts.csv"),
#           row.names = TRUE)
# counts.450k.Houseman <- read.csv(file.path(dataPathCarmona, "cellcounts/450k_minfi-estimateCellCounts.csv"),
#                        row.names = 1)
df.450k.Houseman = gather(cbind("samples" = pData(RGset_target450k)$sampleID, counts.450k.Houseman),
                                "ID" = pData(RGset_target450k)$ID,  
                          celltype, est, -samples)

# load 600 Houseman CpGs from estimateCellCounts.modified()
gr_600CpGs <- granges(Mset_target450k)[rownames(results.450k.Houseman$coefs),]  
findOverlaps(granges(BS), gr_600CpGs) # only 91/600 CpGs overlap between RRBS and 450K sample (600 CpGs)

keepDMRs = methylCC:::pickTargetPositions(targetgRanges = gr_targetbs, targetObject = M_targetbs,
                                         targetCVG = cvg_targetbs, dmpRegions = gr_600CpGs)
x <- findOverlaps(granges(BS), gr_600CpGs) # which of the 600 probes overlap with the RRBS data

coefs <- results.450k.Houseman$coefs[subjectHits(x), ] # dim(coefs) = 91 x 6
counts.RRBS.Houseman <- minfi:::projectCellType(as.data.frame(mcols(keepDMRs$dmpGRanges)), coefs)
counts.RRBS.Houseman <- counts.RRBS.Houseman / rowSums(counts.RRBS.Houseman)
rownames(counts.RRBS.Houseman)[11] <- "D07repA"
rownames(counts.RRBS.Houseman)[12] <- "D07repB"
# write.csv(counts.RRBS.Houseman, file = file.path(dataPathCarmona, "cellcounts/rrbs_minfi-projectCellType.csv"),
#           row.names = TRUE)
# counts.RRBS.Houseman <- read.csv(file.path(dataPathCarmona, "cellcounts/rrbs_minfi-projectCellType.csv"), 
#                                  row.names = 1)
df.RRBS.Houseman = gather(cbind("samples" = rownames(counts.RRBS.Houseman), 
                                counts.RRBS.Houseman), 
                          celltype, est, -samples)
```

Try comparing estimates from only using 91 out of 600 CpGs
```{r Houseman-Carmona-test}
# run estimateCellCounts.modified() on 450k samples (returns 600 Houseman CpGs)
# results.450k.Houseman <- estimateCellCounts.modified(RGset_target450k, returnAll = TRUE)
gr_600CpGs <- granges(Mset_target450k)[rownames(results.450k.Houseman$coefs),]  
findOverlaps(granges(BS), gr_600CpGs) # only 91/600 CpGs overlap between RRBS and 450K sample (600 CpGs)

keepDMRs = methylCC:::pickTargetPositions(targetgRanges = gr_targetbs, targetObject = M_targetbs,
                                         targetCVG = cvg_targetbs, dmpRegions = gr_600CpGs)
x <- findOverlaps(granges(BS), gr_600CpGs) # which of the 600 probes overlap with the RRBS data

coefs <- results.450k.Houseman$coefs[subjectHits(x), ] # dim(coefs) = 91 x 6
counts.RRBS.Houseman <- minfi:::projectCellType(as.data.frame(mcols(keepDMRs$dmpGRanges)), coefs)
counts.RRBS.Houseman = data.frame("Tcell" = rowSums(counts.RRBS.Houseman[,1:3]), counts.RRBS.Houseman[,c(4,5,6)])
counts.RRBS.Houseman <- counts.RRBS.Houseman / rowSums(counts.RRBS.Houseman)
rownames(counts.RRBS.Houseman)[11] <- "D07repA"
rownames(counts.RRBS.Houseman)[12] <- "D07repB"


counts.450k.Houseman <- results.450k.Houseman$counts / rowSums(results.450k.Houseman$counts)
rownames(counts.450k.Houseman) <- pData(RGset_target450k)$sampleID
counts.450k.Houseman = data.frame("Tcell" = rowSums(counts.450k.Houseman[,1:3]), 
                                  counts.450k.Houseman[,c(4,5,6)])
# write.csv(counts.450K.Houseman, file = file.path(dataPathCarmona, "cellCountsEsts/450k_minfi-estimateCellCounts.csv"),
#           row.names = TRUE)
# counts.450k.Houseman <- read.csv(file.path(dataPathCarmona, "cellCountsEsts/450k_minfi-estimateCellCounts.csv"),
#                        row.names = 1)
df.450k.Houseman = gather(cbind("samples" = pData(RGset_target450k)$sampleID, counts.450k.Houseman),
                                #"ID" = pData(RGset_target450k)$ID,  
                          celltype, est, -samples)

# load 600 Houseman CpGs from estimateCellCounts.modified()
gr_600CpGs <- granges(Mset_target450k)[rownames(results.450k.Houseman$coefs),]  
findOverlaps(granges(BS), gr_600CpGs) # only 91/600 CpGs overlap between RRBS and 450K sample (600 CpGs)

keepDMRs = methylCC:::pickTargetPositions(targetgRanges = gr_targetbs, targetObject = M_targetbs,
                                         targetCVG = cvg_targetbs, dmpRegions = gr_600CpGs)
x <- findOverlaps(granges(BS), gr_600CpGs) # which of the 600 probes overlap with the RRBS data

coefs <- results.450k.Houseman$coefs[subjectHits(x), ] # dim(coefs) = 91 x 6
counts.RRBS.Houseman <- minfi:::projectCellType(as.data.frame(mcols(keepDMRs$dmpGRanges)), coefs)
counts.RRBS.Houseman = data.frame("Tcell" = rowSums(counts.RRBS.Houseman[,1:3]), counts.RRBS.Houseman[,c(4,5,6)])
counts.RRBS.Houseman <- counts.RRBS.Houseman / rowSums(counts.RRBS.Houseman)
rownames(counts.RRBS.Houseman)[11] <- "D07repA"
rownames(counts.RRBS.Houseman)[12] <- "D07repB"
# write.csv(counts.RRBS.Houseman, file = file.path(dataPathCarmona, "cellcounts/rrbs_minfi-projectCellType.csv"),
#           row.names = TRUE)
# counts.RRBS.Houseman <- read.csv(file.path(dataPathCarmona, "cellcounts/rrbs_minfi-projectCellType.csv"), 
#                                  row.names = 1)
df.RRBS.Houseman = gather(cbind("samples" = rownames(counts.RRBS.Houseman), 
                                counts.RRBS.Houseman), 
                          celltype, est, -samples)
```



### Create Figure 1

```{r create-Figure1-Carmona}
dfcombined <- full_join(df.450k.Houseman, df.RRBS.Houseman, by = c("samples", "celltype"))
dfcombined$celltype <- factor(dfcombined$celltype, levels = c("Bcell", "Tcell", "Mono", "Gran"))

# colored by cell type
gmat <- ggplot(dfcombined, aes(x=est.x, y = est.y, color = celltype)) + 
  geom_point(size=3) + geom_abline(intercept = 0, slope = 1) + xlim(0,1) + ylim(0,1) +
  xlab("450K platform") + ylab("RRBS platform") + 
  scale_color_discrete(name = "Cell type") + 
  theme(legend.position = "top", legend.justification= "center") + 
  labs(title = "Cell composition estimates from whole blood samples\nmeasured on two platforms (Houseman method)")

pdf(file.path(dataPathCarmona, "figs/Figure1.pdf"), width = 6, height = 6)
print(gmat)
dev.off()
```

# Figure 2

Uses Carmona et al. (2017) data loaded from Figure 1

### Figure 2A

Example genomic region showing the observed methylation 
levels for the same CpGs in the same whole blood 
sample vary depending the platform technology used. 

```{r create-Figure2A}
### Load FlowSorted.Blood.450k object for Figure 2A
# library(FlowSorted.Blood.450k)
Mset <- preprocessIllumina(updateObject(FlowSorted.Blood.450k))
Mset <- mapToGenome(Mset, mergeManifest = FALSE)
pd = as.data.frame(pData(Mset))
p = getBeta(Mset, type = "Illumina")
colnames(p) = pd$Sample_Name = rownames(pd) = gsub("\\+","",pd$Sample_Name)

IDs = c("CD8T","CD4T", "NK","Bcell","Mono","Gran")
p0 = p[,which(pd$CellType %in% IDs)]
pd0 = pd[which(pd$CellType %in% IDs),]

## Use this as the beta-value (b/c outlier at CD8T 105)
p0 = p0[,-which(pd0$Sample_Name == "CD8_105")]
pd0 = pd0[-which(pd0$Sample_Name == "CD8_105"),]
pd0$CellType <- factor(pd0$CellType, levels = IDs)

# Extract chromosome and position information for each probe in 450k array (need this for regions)
gr_train450k <- granges(Mset)
mcols(gr_train450k) <- p0

# merge 450k and RRBS data to same data object to create Gziv plot
dat_all <- merge(gr_target450k, gr_targetbs, all=TRUE)
genome(dat_all) <- "hg19"
gtrack <- GenomeAxisTrack(cex=1.1)

# find regions in 450k data that are on and off
chr <- as.character(seqnames(gr_target450k))
pos <- start(gr_target450k)
cl <- clusterMaker(chr, pos, maxGap = 300)
tab <- regionFinder(rowMeans(as.matrix(mcols(gr_target450k))), chr, pos, cl, cutoff=0.5)

# plot example region: obs DNAm levels in 450k and RRBS samples
ind = 24
tmp = methylCC:::pickTargetPositions(dat_all,  
              dmpRegions = shift(resize(makeGRangesFromDataFrame(tab[ind,]), 
                                        fix = "start", 8e3), 2e3))

dTrack_all <- DataTrack(tmp$probeGRanges, name = "Observed Methylation", type = c("p", "smooth"), 
                        ylim=c(0, 1), cex.title=1.25, cex.axis=1.25, lwd = 3, degree = 1, span = 0.45, 
                        col = gg_color_hue(2), cex= rep(1.5, 1.5))
pdf(file.path(dataPathCarmona, "figs/Fig2A-DMR-24.pdf"), width = 8, height = 5)
plotTracks(list(dTrack_all, gtrack),
           groups = rep(c("450k", "RRBS"), each = 12), 
           main = "Chromosome 6",
           cex.legend = 1.25, cex.main = 1.1)
dev.off()
```


### Figure 2B

Compare beta values (CpGs) in regions methylated or not methylated 
using 450K platform and RRBS (Carmona et al. 2017 data). These 
are the same samples measured on two different platform technologies. 

```{r create-Figure2B}
#### 450k target data
dat.d0.target450k = methylCC:::pickTargetPositions(targetgRanges = gr_target450k, dmpRegions = offMethRegions)
mu0.target450k <- rowMeans(as.data.frame(mcols(dat.d0.target450k$dmpGRanges)))

dat.d1.target450k = methylCC:::pickTargetPositions(targetgRanges = gr_target450k, dmpRegions = onMethRegions)
mu1.target450k <- rowMeans(as.data.frame(mcols(dat.d1.target450k$dmpGRanges)))

#### RRBS target data
dat.d0.targetbs = methylCC:::pickTargetPositions(targetgRanges = gr_targetbs, targetObject = M_targetbs, 
                                      targetCVG = cvg_targetbs, dmpRegions = offMethRegions)
mu0.targetbs <- rowMeans(as.data.frame(mcols(dat.d0.targetbs$dmpGRanges)), na.rm = TRUE)
mu0.targetbs <- mu0.targetbs[mu0.targetbs != 0]

dat.d1.targetbs = methylCC:::pickTargetPositions(targetgRanges = gr_targetbs, targetObject = M_targetbs, 
                                      targetCVG = cvg_targetbs, dmpRegions = onMethRegions)
mu1.targetbs <- rowMeans(as.data.frame(mcols(dat.d1.targetbs$dmpGRanges)), na.rm = TRUE)
mu1.targetbs <- mu1.targetbs[mu1.targetbs != 1]

dat = data.frame("means" = c(mu0.target450k, mu0.targetbs, mu1.target450k, mu1.targetbs), 
                 "Platform" = as.factor(c(rep.int("450k", length(mu0.target450k)), rep.int("RRBS", length(mu0.targetbs)), 
                                          rep.int("450k", length(mu1.target450k)), rep.int("RRBS", length(mu1.targetbs)))), 
                 "Regions" = factor(c(rep("Not methylated", length(c(mu0.target450k, mu0.targetbs))),
                                       rep("Methylated", length(c(mu1.target450k, mu1.targetbs)))), 
                                levels = c("Not methylated", "Methylated")))

p1 <- dat %>% ggplot(aes(x = means, colour = Platform)) + 
    stat_density(aes(colour = Platform, linetype = Regions), geom = "line",position="identity", size = 1.2) + 
    xlab("Methylation") + ylim(c(0, 130)) + xlim(c(0,1)) + 
theme(axis.text.x = element_text(size = rel(1.2)), axis.text.y = element_text(size = rel(1.2)), 
      title = element_text(size = rel(1.3)), panel.background = element_blank(), 
      legend.text = element_text(size = rel(1.2)), 
      legend.position = c(.95, .95),
      legend.justification = c("right", "top"),
      legend.box.just = "right",
      legend.margin = margin(6, 6, 6, 6)) 

pdf(file.path(dataPathCarmona, "figs/Fig2B-platformDependent-density.pdf"), width = 7, height = 5)
print(p1)
dev.off()
```


# Figure 3

Example region with a cell type-spcific Houseman CpG in 450K sample, 
but CpG is not observed in RRBS sample. Using cell type-specific regions
is a solution to this.Uses Carmona et al. (2017) data loaded from Figure 1


```{r create-Figure3}
library(Gviz)
gtrack <- GenomeAxisTrack()

dat_all <- merge(gr_target450k[,4], gr_targetbs[,1], all=TRUE)
genome(dat_all) <- "hg19"

# Combine everything and create individual tracks
expandbump <- 1e2
kk <- subsetByOverlaps(celltypeSpecificDMRegions, gr_600CpGs) + expandbump
ind = 11 
tmp_flowSort = methylCC:::pickTargetPositions(gr_train450k, dmpRegions = kk[ind,])
dTrack_flowSort <- DataTrack(tmp_flowSort$probeGRanges, name = "Observed Methylation", type = c("p", "smooth"), 
                        ylim=c(0, 1), cex.title=1.25, cex.axis=1.25, lwd = 3, degree = 1, span = 0.75,
                        col = gg_color_hue(6), 
                        groups = pd0$CellType)

tmp = methylCC:::pickTargetPositions(dat_all, dmpRegions = kk[ind,])
atrack_CT <- AnnotationTrack(subsetByOverlaps(gr_600CpGs, kk[ind,]), 
                             name = "CpG", stacking = "dense",
                             cex.title=1.1, cex.axis=1.1, rotation.title=0)
atrack_region <- AnnotationTrack(subsetByOverlaps(celltypeSpecificDMRegions, kk[ind,]), 
                                 name = "DMR", stacking = "dense",
                                 cex.title=1.1, cex.axis=1.1, rotation.title=0)

ll <- methylCC:::pickTargetPositions(dat_all, dmpRegions = kk[ind,]- expandbump)
dTrack_all <- DataTrack(ll$probeGRanges, 
                         name = "Observed Methylation", type = c("p"), 
                        ylim=c(0, 1), cex.title=1.25, cex.axis=1.25, lwd = 3,
                        col = gg_color_hue(2), 
                        baseline = c(as.numeric(as.data.frame(mcols(ll$dmpGRanges)))),
                        col.baseline = gg_color_hue(2), lty.baseline = 2, lwd.baseline = 3,
                        groups = rep(c("450k", "RRBS"), each = 1))

pdf(file.path(dataPathCarmona, "figs/Fig3-celltypespecificRegion.pdf"), width = 8, height = 8)
plotTracks(list(dTrack_flowSort, dTrack_all, atrack_CT, atrack_region, gtrack),
           main = "Chromosome 14", 
           from = start(kk[ind,]), to = end(kk[ind,]),
           cex= 1.1, cex.legend = 1.25, cex.main = 1.1)
dev.off()
```



# Figure 4

```{r create-Figure4}
library(methylCC)
data("celltypeSpecificDMRegions")
dat <- as.data.frame(mcols(celltypeSpecificDMRegions))
dat <- as.data.frame(dat[,c(2,1,3,4)])

pdf(file.path(dataPathProject, "Figure4.pdf"), width = 5, height = 6)
imagemat(dat[with(dat, order(-Bcell, -Tcell, -Mono, -Gran)), ], 
         main = "Cell type-specific DNAm profiles\nMethylated (black), Unmethylated (white)", 
         ylab = "Regions (R=212)", xlab = "Cell types", xaxt = "n", 
         cex.axis = 1.3, cex.lab = 1.3, cex.main = 1.3)
axis(1, at=1:4,labels=colnames(dat), 
     cex.axis = 1.3, cex.lab = 1.3, cex.main = 1.3)
dev.off()
```

# Figure 5

### Process Carmona et al. (2017) data

```{r estimateCC-Carmona}
# run methylCC::estimateCC() on 450k samples
est <- methylCC::estimateCC(object = RGset_target450k, verbose = TRUE, epsilon = 0.1,
                             maxIter = 100)
counts.450k.Hicks <- cellcounts(est)
rownames(counts.450k.Hicks) <- colnames(RGset_target450k)
# write.csv(counts.450k.Hicks, file = file.path(dataPathCarmona, "cellcounts/450k_methylCC-estimateCC.csv"),
#           row.names = TRUE)
# counts.450k.Hicks <- read.csv(file.path(dataPathCarmona, "cellcounts/450k_methylCC-estimateCC.csv"), 
#                        row.names = 1)

df.450k.Hicks = gather(cbind("samples" = pData(RGset_target450k)$sampleID, 
                              "ID" = pData(RGset_target450k)$ID, counts.450k.Hicks),
                       celltype, est, -samples, -ID)

# run methylCC::estimateCC() on RRBS data
est <- methylCC::estimateCC(object = BS, verbose = TRUE, epsilon = 0.1, maxIter = 100)
counts.RRBS.Hicks <- cellcounts(est)
# write.csv(counts.RRBS.Hicks,
#           file = file.path(dataPathCarmona, "cellcounts/rrbs_methylCC-estimateCC.csv"),
#           row.names = TRUE)
# counts.RRBS.Hicks <- read.csv(file.path(dataPathCarmona,
#                                         "cellcounts/rrbs_methylCC-estimateCC-old.csv"), 
#                               row.names = 1)

```

### Create Figure 5

```{r create-Figure5-Carmona}
# combine
df.RRBS.Hicks = gather(cbind("samples" = rownames(counts.RRBS.Houseman), 
                             "ID" = pData(BS)$ID, counts.RRBS.Hicks), 
                       celltype, est, -samples, -ID)

dfcombined.450k <- full_join(df.450k.Houseman, df.450k.Hicks, by = c("samples", "celltype"))  
dfcombined.450k$celltype <- factor(dfcombined.450k$celltype, levels = c("Bcell", "Tcell", "Mono", "Gran"))

z.450k <- rbind(data.frame("Method" = rep("Houseman", nrow(df.450k.Houseman)), 
                      df.450k.Houseman), 
           data.frame("Method" = rep("methylCC", nrow(df.450k.Hicks)), 
                      df.450k.Hicks))

z.rrbs <- rbind(data.frame("Method" = rep("Houseman", nrow(df.RRBS.Houseman)), 
                      df.RRBS.Houseman), 
           data.frame("Method" = rep("methylCC", nrow(df.RRBS.Hicks)), 
                      df.RRBS.Hicks))

df.all <- full_join(z.450k, z.rrbs, by = c("Method", "ID", "celltype")) 
df.all$celltype <- factor(df.all$celltype, levels = c("Bcell", "Tcell", "Mono", "Gran"))
df.all$Method <- factor(df.all$Method, levels = c("Houseman", "methylCC"))

gmat <- ggplot(df.all, aes(x=est.x, y = est.y, color = celltype)) + 
  facet_wrap(~Method) + geom_point(size=3) + 
  geom_abline(intercept = 0, slope = 1) + xlim(0,1) + ylim(0,1) +
  xlab("450K platform") + ylab("RRBS platform") + 
  scale_color_discrete(name = "Cell type") + 
  theme(legend.position = "top", legend.justification= "center") + 
  labs(title = "Cell composition estimates from whole blood samples\nmeasured on two platforms")

pdf(file.path(dataPathCarmona, "figs/Figure5.pdf"), width = 8, height = 5)
print(gmat)
dev.off()
```


# Figure S1

Code for Figure S1 can be found at [`scripts/simulationStudy.Rmd`](https://github.com/stephaniehicks/methylCCPaper/blob/master/scripts/simulationStudy.Rmd). 

# Figure S2

### Download Liu et al. (2013) data from GEO 

Run script on JHPCE. 

```{r downloadData-Liu}
# load libraries
library(minfi)
library(bsseq)
library(GenomicRanges)
library(GEOquery)

workingDir_Liu2013 <- "/users/shicks1/projects/methylCC/dataFromLiu2013"
dataPath <- "/users/shicks1/data/GEO/GSE42861"
setwd(dataPath)

message("[dataFromLiu2013-450k] Downloading 450k data.")

# # download all supplemental files (including CEL files)
# dataPath <- "/users/shicks1/data/GEO"
# setwd(dataPath)
# getGEOSuppFiles("GSE42861", makeDirectory = TRUE, baseDir = dataPath)
# untar("GSE42861/GSE42861_RAW.tar", exdir="GSE42861/GSE42861_RAW")
cels <- list.files(file.path(dataPath, "GSE42861_RAW/"), pattern = ".idat")
tmp = paste0("GSE42861_RAW/", unique(gsub(".....idat.gz", "", cels)))

message("Downloading pData.")
gse <- getGEO("GSE42861",GSEMatrix=TRUE, destdir = dataPath)
pd <- pData(gse[[1]])
save(pd, file = file.path(dataPath, "pd_Liu2013.rds"))
rm(gse)

message("Creating RGset.")
RGset <- read.metharray(file.path(dataPath, tmp), 
                         extended = FALSE, verbose = FALSE, force = FALSE)
pData(RGset) <- pd

message("Saving RGset.")
# save RGset object
saveRDS(RGset, file=file.path(dataPath, "RGset_Liu2013.rds"))
```

### Process Liu et al. (2013) data
 
Run script on odyssey. 

```{r estimateCC-Liu}
library(tidyr)
library(dplyr)
library(minfi)
library(methylCC)

workingDir_Liu2013 <- "/users/shicks1/projects/methylCC/dataFromLiu2013"
dataPath <- "/users/shicks1/data/GEO/GSE42861"

# Load 689 samples run on 450K DNA methylation array
RGset <- readRDS(file.path(workingDir_Liu2013, "RGset_Liu2013.rds")) 
RGset <- updateObject(RGset)

# run minfi::estimateCellCounts()
counts.450k.Houseman <- minfi::estimateCellCounts(RGset)
# write.csv(counts.450k.Houseman,
#           file = file.path(dataPathLiu, "cellcounts/450k_minfi-estimateCellCounts.csv"),
#           row.names = TRUE)

Mset <- preprocessIllumina(RGset)
Mset <- mapToGenome(Mset, mergeManifest = FALSE)

message("Save Mset object.")
# saveRDS(Mset, file=file.path(workingDir_Liu2013, "Mset_Liu2013.rds"))
# 
# Mset <- readRDS(file=file.path(workingDir_Liu2013, "Mset_Liu2013.rds"))

pd = pData(Mset)
grObject <- granges(Mset)
betaValues = getBeta(Mset, type = "Illumina")

message("Pick Target Positions")
keepDMRs = methylCC:::pickTargetPositions(targetgRanges=grObject, targetObject = betaValues,
                               dmpRegions=celltypeSpecificDMRegions)$dmpGRanges

zmatSub <- subsetByOverlaps(celltypeSpecificDMRegions, keepDMRs, type = "equal")
rm(betaValues)

ymat <- mcols(keepDMRs); colnames(ymat) <- sampleNames(Mset)
rm(Mset)
n <- ncol(mcols(keepDMRs))
ids <- colnames(mcols(zmatSub))
zmat <- as.matrix(as.data.frame(mcols(zmatSub)))

dat <- list(ymat = ymat, n = n, ids = ids, zmat = zmat,
     grObject = grObject, keepDMRs = keepDMRs)

message("Saving methylCC Processed Data.")
# saveRDS(dat, file=file.path(workingDir_Liu2013, "methylCCProcessed_Liu2013.rds"))


# run methylCC::estimateCC()
dat <- readRDS(file = file.path(workingDir_Liu2013, "methylCCProcessed_Liu2013.rds"))
est <- estimateCC(object = as.matrix(dat$ymat), regionMat = as.matrix(dat$zmat),
                  initParamMethod = "random", epsilon = 0.1, 
                  verbose = TRUE, maxIter = 50)
counts.450k.Hicks <- cellcounts(est)
# write.csv(counts.450k.Hicks,
#           file = file.path(dataPathLiu, "cellcounts/450k_methylCC-estimateCC.csv"),
#           row.names = TRUE)
```


### Create Figure S2 

```{r create-FigureS2-Liu}
# load minfi::estimateCellCounts() estimates
counts.450k.Houseman <- read.csv(file.path(dataPathLiu, "cellCountsEsts/450k_minfi-estimateCellCounts.csv"), 
                                 row.names = 1)
counts.450k.Houseman$Tcell <- rowSums(counts.450k.Houseman[,1:3])
counts.450k.Houseman <- counts.450k.Houseman[,c(7,4:6)]
df.Houseman <- gather(cbind("samples" = rownames(counts.450k.Houseman), counts.450k.Houseman), celltype, est, -samples)

# load methylCC::estimateCC() estimates
counts.450k.Hicks <- read.csv(file.path(dataPathLiu, "cellCountsEsts/450k_methylCC-estimateCC.csv"),  
                              row.names = 1)
df.Hicks <- gather(cbind("samples" = rownames(counts.450k.Hicks), counts.450k.Hicks), celltype, est, -samples)

dfcombined <- full_join(df.Houseman, df.Hicks,  by = c("samples", "celltype"))
dfcombined$celltype <- factor(dfcombined$celltype, levels = c("Bcell", "Tcell", "Mono", "Gran"))

gmat <- dfcombined %>% 
  ggplot(aes(x=est.x, y = est.y, color = celltype)) + 
    geom_abline(intercept = 0, slope = 1) + xlim(0,1) + ylim(0,1) +
    scale_color_discrete(name = "Cell type") +
    geom_point() + theme(legend.position = "top", legend.justification= "center") + 
    xlab("Houseman method") + ylab("methylCC") + 
    labs(title = "Model-based cell composition estimates from\nwhole blood samples (N=689, Li et al. 2013)")

pdf(file.path(dataPathLiu, "figs/FigureS2.pdf"), width = 6, height = 6)
print(gmat)
dev.off()
```


