---
title: "dataFromCarmona whole blood samples processed on RRBS and 450k"
output: html_document
---

```{r}
library(methylCC)
library(dplyr)
library(ggplot2)
library(cowplot)
library(minfi)
library(bsseq)
library(quadprog)
library(GenomicRanges)

workingDir_carmona2017 <- 
  "/users/shicks1/projects/methylCCPaper/case-studies/2017-carmona-wholeblood-rrbs-450"
dataPath <- "/users/shicks1/data/DNAm/2017-carmona-wholeblood-rrbs-450"

estimateCellCounts.modified <- function(rgSet, compositeCellType = "Blood", 
                                         processMethod = "auto", probeSelect = "auto", 
                                cellTypes = c("CD8T", "CD4T", "NK", "Bcell", "Mono", "Gran"),
                                referencePlatform = c("IlluminaHumanMethylation450k", 
                                                      "IlluminaHumanMethylationEPIC", 
                                                      "IlluminaHumanMethylation27k"), 
                                returnAll = FALSE, meanPlot = FALSE, verbose = TRUE, ...) 
{
    minfi:::.isMatrixBackedOrStop(rgSet, "estimateCellCounts")
    minfi:::.isRGOrStop(rgSet)
    rgSet <- as(rgSet, "RGChannelSet")
    referencePlatform <- match.arg(referencePlatform)
    rgPlatform <- sub("IlluminaHumanMethylation", "", annotation(rgSet)[which(names(annotation(rgSet)) == 
        "array")])
    platform <- sub("IlluminaHumanMethylation", "", referencePlatform)
    if ((compositeCellType == "CordBlood") && (!"nRBC" %in% cellTypes)) {
        message("[estimateCellCounts] Consider including 'nRBC' in argument 'cellTypes' for cord blood estimation.\n")
    }
    referencePkg <- sprintf("FlowSorted.%s.%s", compositeCellType, 
        platform)
    subverbose <- max(as.integer(verbose) - 1L, 0L)
    if (!require(referencePkg, character.only = TRUE)) {
        stop(sprintf("Could not find reference data package for compositeCellType '%s' and referencePlatform '%s' (inferred package name is '%s')", 
            compositeCellType, platform, referencePkg))
    }
    data(list = referencePkg)
    referenceRGset <- get(referencePkg)
    if (rgPlatform != platform) {
        rgSet <- convertArray(object = rgSet, outType = referencePlatform, 
            verbose = subverbose)
    }
    if (!"CellType" %in% names(colData(referenceRGset))) {
        stop(sprintf("the reference sorted dataset (in this case '%s') needs to have a phenoData column called 'CellType'"), 
            names(referencePkg))
    }
    if (sum(colnames(rgSet) %in% colnames(referenceRGset)) > 
        0) {
        stop("the sample/column names in the user set must not be in the ", 
            "reference data ")
    }
    if (!all(cellTypes %in% referenceRGset$CellType)) {
        stop(sprintf("all elements of argument 'cellTypes' needs to be part of the reference phenoData columns 'CellType' (containg the following elements: '%s')", 
            paste(unique(referenceRGset$cellType), collapse = "', '")))
    }
    if (length(unique(cellTypes)) < 2) {
        stop("At least 2 cell types must be provided.")
    }
    if ((processMethod == "auto") && (compositeCellType %in% 
        c("Blood", "DLPFC"))) {
        processMethod <- "preprocessQuantile"
    }
    if ((processMethod == "auto") && (!compositeCellType %in% 
        c("Blood", "DLPFC"))) {
        processMethod <- "preprocessNoob"
    }
    processMethod <- get(processMethod)
    if ((probeSelect == "auto") && (compositeCellType == "CordBlood")) {
        probeSelect <- "any"
    }
    if ((probeSelect == "auto") && (compositeCellType != "CordBlood")) {
        probeSelect <- "both"
    }
    if (verbose) {
        message("[estimateCellCounts] Combining user data with reference ", 
            "(flow sorted) data.\n")
    }
    newpd <- DataFrame(sampleNames = c(colnames(rgSet), colnames(referenceRGset)), 
        studyIndex = rep(x = c("user", "reference"), times = c(ncol(rgSet), 
            ncol(referenceRGset))), stringsAsFactors = FALSE)
    referencePd <- colData(referenceRGset)
    combinedRGset <- combineArrays(object1 = rgSet, object2 = referenceRGset, 
        outType = "IlluminaHumanMethylation450k")
    colData(combinedRGset) <- newpd
    colnames(combinedRGset) <- newpd$sampleNames
    rm(referenceRGset)
    if (verbose) {
        message("[estimateCellCounts] Processing user and reference data ", 
            "together.\n")
    }
    if (compositeCellType == "CordBlood") {
        combinedMset <- processMethod(combinedRGset, verbose = subverbose)
        compTable <- get(paste0(referencePkg, ".compTable"))
        combinedMset <- combinedMset[which(rownames(combinedMset) %in% 
            rownames(compTable)), ]
    }
    else {
        combinedMset <- processMethod(combinedRGset)
    }
    rm(combinedRGset)
    referenceMset <- combinedMset[, combinedMset$studyIndex == 
        "reference"]
    colData(referenceMset) <- as(referencePd, "DataFrame")
    mSet <- combinedMset[, combinedMset$studyIndex == "user"]
    colData(mSet) <- as(colData(rgSet), "DataFrame")
    rm(combinedMset)
    if (verbose) {
        message("[estimateCellCounts] Picking probes for composition ", 
            "estimation.\n")
    }
    compData <- minfi:::pickCompProbes(mSet = referenceMset, 
                                       cellTypes = cellTypes, 
                                       compositeCellType = compositeCellType, 
                                       probeSelect = probeSelect)
    coefs <- compData$coefEsts
    # rm(referenceMset)
    if (verbose) 
        message("[estimateCellCounts] Estimating composition.\n")
    counts <- minfi:::projectCellType(getBeta(mSet)[rownames(coefs), 
        ], coefs)
    rownames(counts) <- colnames(rgSet)
    if (meanPlot) {
        smeans <- compData$sampleMeans
        smeans <- smeans[order(names(smeans))]
        sampleMeans <- c(colMeans2(x = getBeta(mSet), rows = match(rownames(coefs), 
            rownames(mSet))), smeans)
        sampleColors <- c(rep(1, ncol(mSet)), 1 + as.numeric(factor(names(smeans))))
        plot(sampleMeans, pch = 21, bg = sampleColors)
        legend("bottomleft", c("blood", 
                               levels(factor(names(smeans)))), 
               col = 1:7, pch = 15)
    }
    if (returnAll) {
        return(list(counts = counts, coefs = coefs, compTable = compData$compTable,
             normalizedData = mSet, normalizedMsetReference = referenceMset))
    }
    else {
        counts
    }
}

```


# Load 450K and RRBS data from Carmona

## Load target data: 450K samples 

```{r}
load(file.path(dataPath, "2017-carmona-mset450-krawall.RData"))
mset450krawall <- updateObject(mset450krawall)
RGset_target450k <- mset450krawall
rm(mset450krawall)

# This is needed to run the estimateCellCounts() function
colData(RGset_target450k)$Sentrix_Position <- as.character(colData(RGset_target450k)$Sentrix_Position)

# add ID to pData()
library(stringr)
newnames <- sapply(pData(RGset_target450k)$sampleID, function(x){
  str_split(x, pattern = "_")[[1]][1] })
pData(RGset_target450k)$ID <- newnames
RGset_target450k <- RGset_target450k[, !(pData(RGset_target450k)$ID %in% 
                                           c("D07repB", "D07repC"))]

# create target450k datasets
Mset_target450k <- preprocessIllumina(RGset_target450k) 
Mset_target450k <- mapToGenome(Mset_target450k, mergeManifest = FALSE)

pd_target450k = as.data.frame(pData(Mset_target450k))
p_target450k = getBeta(Mset_target450k, type = "Illumina")
gr_target450k <- granges(Mset_target450k)

mcols(gr_target450k) <- p_target450k
```

## Load target data: RRBS samples 

```{r}
# load(file.path(dataPath, "combined.data.Best12.RData"))
# dat <- combined.data.Best12 # list of three data.frames with 6million rows each (DO NOT print by itself!)
# rm(combined.data.Best12)
# 
# dat$methyl <- round(((0.01) * dat$percentmethyl[,-c(1,2)]) * dat$numberreads[,-c(1,2)] )
# dat$methyl[is.na(dat$methyl)] <- 0 # remove NAs
# dat$numberreads[is.na(dat$numberreads)] <- 0 # remove NAs
# 
# newnames <- sapply(colnames(dat$methyl), function(x){
#   paste(str_c(str_split(x, pattern = "\\.")[[1]][2:4], collapse = "_"), "R1.fastq", sep="_") })
# newnames[11:12] <- c("D07repA_RRBS", "D07repB_RRBS")
# colnames(dat$numberreads)[-c(1,2)] = colnames(dat$methyl) = newnames
# 
# message("[2017-carmona-wholeblood-rrbs-450] Creating BSseq object.")
# # create BSeq Object
# BS <- BSseq(pos = dat$numberreads$start,
#             chr = dat$numberreads$chr,
#             M = as.matrix(dat$methyl),
#             Cov = as.matrix(dat$numberreads[,-c(1,2)]))
# 
# BS <- BS[, !(colnames(dat$methyl) %in% c("B632_CGATGT_L006_R1.fastq"))]
# 
# # add pData() object; create ID column
# newIDs <- sapply(colnames(dat$methyl[, !(colnames(dat$methyl)
#                  %in% c("B632_CGATGT_L006_R1.fastq"))]), function(x){
#                    str_split(x, pattern = "_")[[1]][1] })
# pData(BS) <- DataFrame("sampleID" = colnames(dat$methyl[, !(colnames(dat$methyl)
#                                                   %in% c("B632_CGATGT_L006_R1.fastq"))]),
#                        ID = newIDs)
# sampleNames(BS) <- pData(BS)$sampleID
# rm(dat)
# BS <- updateObject(BS)
# saveRDS(BS, file=file.path(dataPath, "2017-carmona-bsseq-krawall.RDS"))

BS <- readRDS(file.path(dataPath, "2017-carmona-bsseq-krawall.RDS"))
BS <- BS[, !(pData(BS)$ID %in%  c("D07repB"))]

cvg_targetbs <- as.data.frame(getBSseq(BS, type = "Cov"))
M_targetbs <- as.data.frame(bsseq::getBSseq(BS, type = "M"))
p_targetbs <- as.data.frame(bsseq::getMeth(BS, type="raw")) # meth / cvg
p_targetbs <- as.data.frame(p_targetbs)
pd_targetbs <- pData(BS)
gr_targetbs <- granges(BS)
mcols(gr_targetbs) <- p_targetbs

# # length(gr_target450k) = 485512
# # length(gr_targetbs) = 6823620
# # How many CpGs overlap?
# intersectProbes = subsetByOverlaps(gr_target450k, gr_targetbs)
# intersectProbes # length(intersectProbes) = 142002

```



# Run `estimateCellCounts.modified ()` (Houseman model) on `RGset_target450k` 

```{r}
# results.450k.Houseman <-
#   estimateCellCounts.modified(RGset_target450k,
#                               returnAll = TRUE,
#                               celltypes = IDs)
# counts.450k.Houseman <- results.450k.Houseman$counts
# rownames(counts.450k.Houseman) <- pData(RGset_target450k)$sampleID
# 
# write.csv(counts.450k.Houseman,
#           file = file.path(workingDir_carmona2017,
#                 "ests/450k-minfi-estimateCellCounts.csv"),
#           row.names = TRUE)

counts.450k.Houseman <- 
  read.csv(file.path(workingDir_carmona2017, 
                     "ests/450k-minfi-estimateCellCounts.csv"), 
           row.names = 1)
counts.450k.Houseman

df.450k.Houseman <- 
  tidyr::gather(cbind("samples" = pData(RGset_target450k)$sampleID,
                                       "ID" = pData(RGset_target450k)$ID,
                                "rep" = c(rep(TRUE, 1), rep(FALSE, 9)), 
                                as.data.frame(counts.450k.Houseman)), 
                          celltype, est, -samples, -rep, -ID) %>% 
  select(-rep)
```


# Run `methylCC::estimatecc()` on `RGset_target450k` 

Including DMRs
```{r}
# output <- find_dmrs(gr_target = NULL,
#                     include_cpgs = TRUE, include_dmrs = TRUE)
# 
# # run methylCC::estimatecc
# set.seed(12345)
# est <- estimatecc(object = RGset_target450k,
#                   find_dmrs_object = output,
#                   verbose = TRUE, epsilon = 0.01,
#                   max_iter = 100, init_param_method = "random")
# counts.450k.Hicks <- cell_counts(est)
# rownames(counts.450k.Hicks) <- colnames(RGset_target450k)
# 
# write.csv(counts.450k.Hicks,
#           file = file.path(workingDir_carmona2017,
#                 "ests/450k-methylCC-estimatecc.csv"),
#           row.names = TRUE)

counts.450k.Hicks <- 
  read.csv(file.path(workingDir_carmona2017, 
                     "ests/450k-methylCC-estimatecc.csv"), 
           row.names = 1)

df.450k.Hicks = tidyr::gather(cbind("samples" = pData(RGset_target450k)$sampleID, 
                              "ID" = pData(RGset_target450k)$ID, counts.450k.Hicks),
                       celltype, est, -samples, -ID)

dfcombined <- dplyr::full_join(df.450k.Houseman, df.450k.Hicks, by = c("samples", "celltype"))
dfcombined$celltype <- factor(dfcombined$celltype, levels = c("CD8T", "CD4T", "NK", "Bcell", "Mono", "Gran"))

dfcombined %>% 
  ggplot(aes(x=est.x, y = est.y, colour = celltype)) +
    geom_abline(intercept = 0, slope = 1) + xlim(0,1) + ylim(0,1) +
     geom_point() +
    xlab("Using minfi::estimateCellCounts()") + ylab("Using methylCC::estimateCC()") +
    labs(title = "Comparing cell composition estimates (450K data only)")

```


Now ONLY CpGs (for reviewer)
```{r}
output_onlyCpGs <- find_dmrs(gr_target = NULL,
                    include_cpgs = TRUE, include_dmrs = FALSE)

# run methylCC::estimatecc
set.seed(12345)
est <- estimatecc(object = RGset_target450k,
                  find_dmrs_object = output_onlyCpGs,
                  verbose = TRUE, epsilon = 0.01,
                  max_iter = 100, init_param_method = "random")
counts.450k.Hicks.onlyCpGs <- cell_counts(est)
rownames(counts.450k.Hicks.onlyCpGs) <- colnames(RGset_target450k)

write.csv(counts.450k.Hicks.onlyCpGs,
          file = file.path(workingDir_carmona2017,
                "ests/450k-methylCC-estimatecc-onlyCpGs.csv"),
          row.names = TRUE)

counts.450k.Hicks.onlyCpGs <- 
  read.csv(file.path(workingDir_carmona2017, 
                     "ests/450k-methylCC-estimatecc-onlyCpGs.csv"), 
           row.names = 1)

df.450k.Hicks.onlyCpGs = tidyr::gather(cbind("samples" = pData(RGset_target450k)$sampleID, 
                              "ID" = pData(RGset_target450k)$ID, counts.450k.Hicks.onlyCpGs),
                       celltype, est, -samples, -ID)
```

# Run `minfi:::projectCellType()` (Houseman model)  on RRBS data

```{r}
output_nogr <- find_dmrs(gr_target = NULL, 
                         include_cpgs = TRUE, include_dmrs = TRUE)

dim(output_nogr$cpgs_houseman) # 600 CpGs (no intersection with RRBS data)

subsetByOverlaps(gr_targetbs, output_nogr$regions_houseman)
# 102 CpGs out of 600 CpGs overlap

sub_cpgs <- names(subsetByOverlaps(output_nogr$regions_houseman, gr_targetbs))

counts.RRBS.Houseman <- 
  minfi:::projectCellType(as.data.frame(mcols(
              subsetByOverlaps(gr_targetbs, output_nogr$regions_houseman))), 
              output_nogr$cpgs_houseman[sub_cpgs, ])
rownames(counts.RRBS.Houseman) <- pData(BS)$ID

# combine
df.RRBS.Houseman = tidyr::gather(cbind("samples" = pData(BS)$sampleID, 
                                       "ID" = pData(BS)$ID, 
                                       as.data.frame(counts.RRBS.Houseman)), 
                       celltype, est, -samples, -ID)
```


# Run `minfi:::projectCellType()` (Houseman model modified)  on RRBS data

```{r}
output <- find_dmrs(gr_target = gr_targetbs,
                    include_cpgs = TRUE, include_dmrs = TRUE)

dim(output$cpgs_houseman) # 587 CpGs (with intersection with RRBS data)

subsetByOverlaps(gr_targetbs, output$regions_houseman)
# 587 CpGs out of 587 CpGs overlap

subsetByOverlaps(output_nogr$regions_houseman, output$regions_houseman) # 102/102

counts.RRBS.Houseman.mod <- 
  minfi:::projectCellType(as.data.frame(mcols( 
              subsetByOverlaps(gr_targetbs, 
                               output$regions_houseman))), 
              output$cpgs_houseman)
rownames(counts.RRBS.Houseman.mod) <- pData(BS)$ID

# combine
df.RRBS.Houseman.mod = tidyr::gather(cbind("samples" = pData(BS)$sampleID, 
                                       "ID" = pData(BS)$ID, 
                                       as.data.frame(counts.RRBS.Houseman.mod)), 
                       celltype, est, -samples, -ID)
```


```{r}
pdf(file.path(workingDir_carmona2017, 
              "figs/fig-carmona2017-houseman-coefs.pdf"), width = 10, height = 6)
par(mfrow=c(1,3))
image(t(output_nogr$cpgs_houseman), xaxt="n", main = "Houseman method:\ntop 600 CpGs")
axis(side=1, at=seq(0, 1, by =.2), labels=colnames(output_nogr$zmat))
image(t(output_nogr$cpgs_houseman[sub_cpgs, ]), xaxt="n", main = "Houseman method:\n 102 out of top 600 CpGs overlapping")
axis(side=1, at=seq(0, 1, by =.2), labels=colnames(output_nogr$zmat))
image(t(output$cpgs_houseman), xaxt="n", main = "Houseman method (modified):\ntop 587 CpGs")
axis(side=1, at=seq(0, 1, by =.2), labels=colnames(output$zmat))
dev.off()
```


# Run `methylCC::estimatecc()` on a `BS` (RRBS data)

Including DMRs
```{r}
# output_nogr <- find_dmrs(gr_target = NULL, 
#                          include_cpgs = TRUE, include_dmrs = TRUE)
# 
# table(output_nogr$regions_all$cellType, output_nogr$regions_all$L,
#       output_nogr$regions_all$dmr_status, output_nogr$regions_all$status)
# 
# p1 <- as.data.frame(mcols(output_nogr$regions_all)) %>%
#   ggplot(aes(x = dm, y = -log(p.value),
#              color = dmr_status)) +
#   geom_point() +
#   geom_hline(yintercept=-log(1e-11)) +
#   geom_hline(yintercept=-log(1e-08))
# 
# p2 <- as.data.frame(mcols(output_nogr$regions_all)) %>%
#   ggplot(aes(x = dm, y = -log(p.value),
#              color = cellType)) +
#   geom_point() +
#   geom_hline(yintercept=-log(1e-11)) +
#   geom_hline(yintercept=-log(1e-08))
# 
# plot_grid(p1, p2)
# 
# # run methylCC::estimatecc
# set.seed(12345)
# est_nogr <- estimatecc(object = BS,
#                   find_dmrs_object = output_nogr,
#                   verbose = TRUE, epsilon = 0.01,
#                   max_iter = 100, init_param_method = "random")
# 
# # [estimatecc] BSseq object contained
# #                 141 out of 600 celltype-specific regions.
# # [estimatecc] Starting parameter estimation using 141 regions.
# 
# counts.RRBS.Hicks <- cell_counts(est_nogr)
# rownames(counts.RRBS.Hicks) <- pData(BS)$ID
# 
# write.csv(counts.RRBS.Hicks,
#           file = file.path(workingDir_carmona2017, "ests/rrbs-methylCC-estimatecc.csv"),
#           row.names = TRUE)

counts.RRBS.Hicks <- read.csv(file.path(workingDir_carmona2017,
                                        "ests/rrbs-methylCC-estimatecc.csv"), 
                              row.names = 1)

df.RRBS.Hicks = tidyr::gather(cbind("samples" = pData(BS)$sampleID, 
                             "ID" = pData(BS)$ID, counts.RRBS.Hicks), 
                       celltype, est, -samples, -ID)

```

Now ONLY CpGs (for reviewer)
```{r}
output_onlyCpGs <- find_dmrs(gr_target = NULL,
                    include_cpgs = TRUE, include_dmrs = FALSE)

# run methylCC::estimatecc
set.seed(12345)
est <- estimatecc(object = BS,
                  find_dmrs_object = output_onlyCpGs,
                  verbose = TRUE, epsilon = 0.01,
                  max_iter = 100, init_param_method = "random")
# [estimatecc] BSseq object contained
#                 101 out of 600 celltype-specific regions.
# [estimatecc] Starting parameter estimation using 101 regions.

counts.RRBS.Hicks.onlyCpGs <- cell_counts(est)
rownames(counts.RRBS.Hicks.onlyCpGs) <- pData(BS)$ID

write.csv(counts.RRBS.Hicks.onlyCpGs,
          file = file.path(workingDir_carmona2017,
                "ests/rrbs-methylCC-estimatecc-mod-onlyCpGs.csv"),
          row.names = TRUE)

counts.RRBS.Hicks.onlyCpGs <- 
  read.csv(file.path(workingDir_carmona2017, 
                     "ests/rrbs-methylCC-estimatecc-mod-onlyCpGs.csv"), 
           row.names = 1)

df.RRBS.Hicks.onlyCpGs = tidyr::gather(cbind("samples" = pData(BS)$sampleID, 
                             "ID" = pData(BS)$ID, counts.RRBS.Hicks.onlyCpGs), 
                       celltype, est, -samples, -ID)
```

Including DMRs, but also including `gr_target`
```{r}
# output <- find_dmrs(gr_target = gr_targetbs,
#                     include_cpgs = TRUE, include_dmrs = TRUE)#,
#                     # cpg_up_dm_cutoff = -0.1,
#                     # cpg_down_dm_cutoff = 0.1)
# 
# table(output$regions_all$cellType, output$regions_all$L,
#       output$regions_all$dmr_status, output$regions_all$status)
# 
# p1 <- as.data.frame(mcols(output$regions_all)) %>%
#   ggplot(aes(x = dm, y = -log(p.value),
#              color = dmr_status)) +
#   geom_point() +
#   geom_hline(yintercept=-log(1e-11)) +
#   geom_hline(yintercept=-log(1e-08))
# 
# 
# p2 <- as.data.frame(mcols(output$regions_all)) %>%
#   ggplot(aes(x = dm, y = -log(p.value),
#              color = cellType)) +
#   geom_point() +
#   geom_hline(yintercept=-log(1e-11)) +
#   geom_hline(yintercept=-log(1e-08))
# 
# plot_grid(p1, p2)
# 
# # run methylCC::estimatecc
# set.seed(12345)
# est <- estimatecc(object = BS,
#                   find_dmrs_object = output,
#                   verbose = TRUE, epsilon = 0.01,
#                   max_iter = 100, init_param_method = "random")
# # [estimatecc] BSseq object contained
# #                 520 out of 520 celltype-specific regions.
# # [estimatecc] Starting parameter estimation using 520 regions.
# 
# counts.RRBS.Hicks.mod <- cell_counts(est)
# rownames(counts.RRBS.Hicks.mod) <- pData(BS)$ID
# 
# write.csv(counts.RRBS.Hicks.mod,
#           file = file.path(workingDir_carmona2017, "ests/rrbs-methylCC-estimatecc-mod.csv"),
#           row.names = TRUE)
# 
counts.RRBS.Hicks.mod <- read.csv(file.path(workingDir_carmona2017,
                                        "ests/rrbs-methylCC-estimatecc-mod.csv"),
                              row.names = 1)

df.RRBS.Hicks.mod = tidyr::gather(cbind("samples" = pData(BS)$sampleID, 
                             "ID" = pData(BS)$ID, counts.RRBS.Hicks.mod), 
                       celltype, est, -samples, -ID)

```



# Run `methylCC::estimatecc()` on a `BS` (RRBS data) but using regions identified with WGBS

```{r}
output_nogr_wgbs <- find_dmrs_wgbs(num_regions = 100, dmr_up_cutoff = 1, 
                              dmr_down_cutoff = 1)

# run methylCC::estimatecc
set.seed(12345)
est <- estimatecc(object = BS,
                  find_dmrs_object = output_nogr_wgbs,
                  verbose = TRUE, epsilon = 0.01,
                  max_iter = 100, init_param_method = "random")
# [estimatecc] BSseq object contained
#                 126 out of 435 celltype-specific regions.
# [estimatecc] Starting parameter estimation using 520 regions.

counts.RRBS.Hicks.wgbs <- cell_counts(est)
rownames(counts.RRBS.Hicks.wgbs) <- pData(BS)$ID

# write.csv(counts.RRBS.Hicks.wgbs,
#           file = file.path(workingDir_carmona2017, "ests/rrbs-methylCC-estimatecc-wgbsregions.csv"),
#           row.names = TRUE)
# 
# counts.RRBS.Hicks.wgbs <- read.csv(file.path(workingDir_carmona2017,
#                                         "ests/rrbs-methylCC-estimatecc-mod.csv"),
#                               row.names = 1)

df.RRBS.Hicks.wgbs = tidyr::gather(cbind("samples" = pData(BS)$sampleID, 
                             "ID" = pData(BS)$ID, counts.RRBS.Hicks.wgbs), 
                       celltype, est, -samples, -ID)

```




# Combine everything together

```{r}
z.450k <- rbind(data.frame("Method" = rep("Houseman", nrow(df.450k.Houseman)), 
                      df.450k.Houseman),
                data.frame("Method" = rep("Houseman (modified)", nrow(df.450k.Houseman)), 
                      df.450k.Houseman),
                data.frame("Method" = rep("methylCC", nrow(df.450k.Hicks)), 
                      df.450k.Hicks),
                data.frame("Method" = rep("methylCC (modified)", nrow(df.450k.Hicks)), 
                      df.450k.Hicks), 
                data.frame("Method" = rep("methylCC (CpGs only)", nrow(df.450k.Hicks)), 
                      df.450k.Hicks.onlyCpGs), 
                data.frame("Method" = rep("methylCC (using 450K regions)", nrow(df.450k.Hicks)), 
                      df.450k.Hicks), 
                 data.frame("Method" = rep("methylCC (using WGBS regions)", nrow(df.450k.Hicks)), 
                      df.450k.Hicks))

z.rrbs <- rbind(data.frame("Method" = rep("Houseman", nrow(df.RRBS.Houseman)), 
                      df.RRBS.Houseman),
                data.frame("Method" = rep("Houseman (modified)", nrow(df.RRBS.Houseman.mod)), 
                      df.RRBS.Houseman.mod), 
                data.frame("Method" = rep("methylCC", nrow(df.RRBS.Hicks)), 
                      df.RRBS.Hicks), 
                data.frame("Method" = rep("methylCC (modified)", nrow(df.RRBS.Hicks.mod)), 
                      df.RRBS.Hicks.mod),
                data.frame("Method" = rep("methylCC (CpGs only)", nrow(df.RRBS.Hicks.mod)), 
                      df.RRBS.Hicks.onlyCpGs),
                data.frame("Method" = rep("methylCC (using 450K regions)", nrow(df.RRBS.Hicks.mod)), 
                      df.RRBS.Hicks), 
                data.frame("Method" = rep("methylCC (using WGBS regions)", nrow(df.RRBS.Hicks.wgbs)), 
                      df.RRBS.Hicks.wgbs))

df.all <- full_join(z.450k, z.rrbs, by = c("Method", "ID", "celltype"))
colnames(df.all)[colnames(df.all) %in% c("est.x", "est.y")] <- 
  c("450K platform", "RRBS platform")
df.all$celltype <- factor(df.all$celltype)

dat_text <- as_tibble(df.all) %>% 
  filter(!(Method %in% c("methylCC (modified)"))) %>% 
  group_by(Method) %>% 
  mutate(sq_diff = (`450K platform` - `RRBS platform`)^2) %>% 
  summarize(rmse=sqrt(mean(sq_diff))) %>% 
  mutate(label = paste0("RMSE: ", round(rmse,4))) %>% 
  mutate(x = rep(0, 6)) %>% 
  mutate(y = rep(0.75,6))
dat_text

gmat <-
  df.all %>% 
  filter(!(Method %in% c("methylCC (modified)"))) %>% 
  ggplot(aes(x=`450K platform`, y = `RRBS platform`)) + 
  facet_wrap(~ Method) + geom_point(size=3, aes(color = celltype)) + 
  geom_abline(intercept = 0, slope = 1) + xlim(0,1) + ylim(0,1) +
  xlab("450K platform") + ylab("RRBS platform") + 
  theme(legend.position = "top", legend.justification= "center") + 
  labs(title = "Cell composition estimates from whole blood samples\nmeasured on two platforms") + 
  scale_color_discrete(name = "Cell type", 
                       labels = c("B-cells", "CD4 T-cells", "CD8 T-cells", 
                                  "Granulocytes","Monocytes", "Natural killer cells"))
gmat <- gmat + geom_text(
  data    = dat_text,
  mapping = aes(x = x, y = y, label = label),
  hjust   = -0.1,
  vjust   = -1,
  size = 5, 
  fontface=2
)

pdf(file.path(workingDir_carmona2017, 
              "figs/fig-carmona2017-houseman-ests-all.pdf"), width = 12, height = 8)
print(gmat)
dev.off()



# 450K vs WGBS DMR comparison
dat_text <- as_tibble(df.all) %>% 
  group_by(Method) %>% 
  mutate(sq_diff = (`450K platform` - `RRBS platform`)^2) %>% 
  summarize(rmse=sqrt(mean(sq_diff))) %>% 
  mutate(label = paste0("RMSE: ", round(rmse,4))) %>% 
  mutate(x = rep(0, 7)) %>% 
  mutate(y = rep(0.75,7)) %>% 
  filter(!(Method %in% c("methylCC (modified)", "methylCC", 
                         "methylCC (CpGs only)", "Houseman (modified)")))
dat_text

gmat <-
  df.all %>% 
  filter(!(Method %in% c("methylCC (modified)", "methylCC", 
                         "methylCC (CpGs only)", "Houseman (modified)"))) %>%
  ggplot(aes(x=`450K platform`, y = `RRBS platform`)) + 
  facet_grid(~ Method) + geom_point(size=3, aes(color = celltype)) + 
  geom_abline(intercept = 0, slope = 1) + xlim(0,1) + ylim(0,1) +
  xlab("450K platform") + ylab("RRBS platform") + 
  theme(legend.position = "top", legend.justification= "center") + 
  labs(title = "Cell composition estimates from whole blood samples\nmeasured on two platforms") + 
  scale_color_discrete(name = "Cell type", 
                       labels = c("B-cells", "CD4 T-cells", "CD8 T-cells", 
                                  "Granulocytes","Monocytes", "Natural killer cells"))
gmat <- gmat + geom_text(
  data    = dat_text,
  mapping = aes(x = x, y = y, label = label),
  hjust   = -0.1,
  vjust   = -1,
  size = 5, 
  fontface=2
)

pdf(file.path(workingDir_carmona2017, 
              "figs/fig-carmona2017-houseman-ests-all-3.pdf"), width = 10, height = 5)
print(gmat)
dev.off()


# CpG vs Region comparison
dat_text <- as_tibble(df.all) %>% 
  group_by(Method) %>% 
  mutate(sq_diff = (`450K platform` - `RRBS platform`)^2) %>% 
  summarize(rmse=sqrt(mean(sq_diff))) %>% 
  mutate(label = paste0("RMSE: ", round(rmse,4))) %>% 
  mutate(x = rep(0, 7)) %>% 
  mutate(y = rep(0.75,7)) %>% 
  filter(!(Method %in% c("methylCC (modified)", "methylCC (using 450K regions)", 
                         "methylCC (using WGBS regions)", "Houseman (modified)")))
dat_text

gmat <-
  df.all %>% 
  filter(!(Method %in% c("methylCC (modified)", "methylCC (using 450K regions)", 
                         "methylCC (using WGBS regions)", "Houseman (modified)"))) %>%
  ggplot(aes(x=`450K platform`, y = `RRBS platform`)) + 
  facet_grid(~ Method) + geom_point(size=3, aes(color = celltype)) + 
  geom_abline(intercept = 0, slope = 1) + xlim(0,1) + ylim(0,1) +
  xlab("450K platform") + ylab("RRBS platform") + 
  theme(legend.position = "top", legend.justification= "center") + 
  labs(title = "Cell composition estimates from whole blood samples\nmeasured on two platforms") + 
  scale_color_discrete(name = "Cell type", 
                       labels = c("B-cells", "CD4 T-cells", "CD8 T-cells", 
                                  "Granulocytes","Monocytes", "Natural killer cells"))
gmat <- gmat + geom_text(
  data    = dat_text,
  mapping = aes(x = x, y = y, label = label),
  hjust   = -0.1,
  vjust   = -1,
  size = 5, 
  fontface=2
)

pdf(file.path(workingDir_carmona2017, 
              "figs/fig-carmona2017-houseman-ests-all-3-cpg-comparison.pdf"), width = 10, height = 5)
print(gmat)
dev.off()


# Figure 4
dat_text <- as_tibble(df.all) %>% 
  group_by(Method) %>% 
  mutate(sq_diff = (`450K platform` - `RRBS platform`)^2) %>% 
  summarize(rmse=sqrt(mean(sq_diff))) %>% 
  mutate(label = paste0("RMSE: ", round(rmse,4))) %>% 
  mutate(x = rep(0, 7)) %>% 
  mutate(y = rep(0.75,7)) %>% 
  filter(!(Method %in% c("methylCC (modified)", "methylCC (using 450K regions)", 
                         "methylCC (using WGBS regions)", "Houseman (modified)", 
                         "methylCC (CpGs only)")))
dat_text

gmat <-
  df.all %>% 
  filter(!(Method %in% c("methylCC (modified)", "methylCC (using 450K regions)", 
                         "methylCC (using WGBS regions)", "Houseman (modified)", 
                         "methylCC (CpGs only)"))) %>%
  ggplot(aes(x=`450K platform`, y = `RRBS platform`)) + 
  facet_grid(~ Method) + geom_point(size=3, aes(color = celltype)) + 
  geom_abline(intercept = 0, slope = 1) + xlim(0,1) + ylim(0,1) +
  xlab("450K platform") + ylab("RRBS platform") + 
  theme(legend.position = "top", legend.justification= "center") + 
  labs(title = "Cell composition estimates from whole blood samples\nmeasured on two platforms") + 
  scale_color_discrete(name = "Cell type", 
                       labels = c("B-cells", "CD4 T-cells", "CD8 T-cells", 
                                  "Granulocytes","Monocytes", "Natural killer cells"))
gmat <- gmat + geom_text(
  data    = dat_text,
  mapping = aes(x = x, y = y, label = label),
  hjust   = -0.1,
  vjust   = -1,
  size = 5, 
  fontface=2
)


pdf(file.path(workingDir_carmona2017, 
              "figs/fig-carmona2017-houseman-ests-all-2.pdf"), width = 8, height = 5)
print(gmat)
dev.off()



gmat <-
  df.all %>% 
  filter(Method %in% c("Houseman")) %>%
  ggplot(aes(x=`450K platform`, y = `RRBS platform`)) + 
  geom_point(size=3, aes(color = celltype)) + 
  geom_abline(intercept = 0, slope = 1) + xlim(0,1) + ylim(0,1) +
  xlab("450K platform") + ylab("RRBS platform") + 
  theme(legend.position = "top", legend.justification= "center") + 
  labs(title = "Cell composition estimates from whole blood samples\nmeasured on two platforms (Houseman method)") + 
  scale_color_discrete(name = "Cell type", 
                       labels = c("B-cells", "CD4 T-cells", "CD8 T-cells", 
                                  "Granulocytes","Monocytes", "Natural killer cells"))

pdf(file.path(workingDir_carmona2017, 
              "figs/fig-carmona2017-houseman-ests-all-1.pdf"), width = 6, height = 6)
print(gmat)
dev.off()

```


# Figure 2

Uses Carmona et al. (2017) data loaded above

### Figure 2A

Example genomic region showing the observed methylation 
levels for the same CpGs in the same whole blood 
sample vary depending the platform technology used. 

```{r create-Figure2A}
library(Gviz)
library(FlowSorted.Blood.450k)

gg_color_hue <- function(n) {
  hues = seq(15, 375, length=n+1)
  hcl(h=hues, l=65, c=100)[1:n]
}

### Load FlowSorted.Blood.450k object for Figure 2A
Mset <- preprocessIllumina(updateObject(FlowSorted.Blood.450k))
Mset <- mapToGenome(Mset, mergeManifest = FALSE)
pd = as.data.frame(pData(Mset))
p = getBeta(Mset, type = "Illumina")
colnames(p) = pd$Sample_Name = rownames(pd) = gsub("\\+","",pd$Sample_Name)

IDs = c("CD8T","CD4T", "NK","Bcell","Mono","Gran")
p0 = p[,which(pd$CellType %in% IDs)]
pd0 = pd[which(pd$CellType %in% IDs),]

## Use this as the beta-value (b/c outlier at CD8T 105)
p0 = p0[,-which(pd0$Sample_Name == "CD8_105")]
pd0 = pd0[-which(pd0$Sample_Name == "CD8_105"),]
pd0$CellType <- factor(pd0$CellType, levels = IDs)

# Extract chromosome and position information for each probe in 450k array (need this for regions)
gr_train450k <- granges(Mset)
mcols(gr_train450k) <- p0

# merge 450k and RRBS data to same data object to create Gziv plot
dat_all <- merge(gr_target450k, gr_targetbs, all=TRUE)
genome(dat_all) <- "hg19"
gtrack <- GenomeAxisTrack(cex=1.1)

# find regions in 450k data that are on and off
chr <- as.character(seqnames(gr_target450k))
pos <- start(gr_target450k)
cl <- clusterMaker(chr, pos, maxGap = 300)
tab <- regionFinder(rowMeans(as.matrix(mcols(gr_target450k))), chr, pos, cl, cutoff=0.5)

# plot example region: obs DNAm levels in 450k and RRBS samples
ind <- 24
tmp <-  methylCC:::.pick_target_positions(dat_all,  
                            dmp_regions = shift(resize(makeGRangesFromDataFrame(tab[ind,]), 
                                                       fix = "start", 8e3), 2e3))

dTrack_all <- DataTrack(tmp$probeGRanges, name = "Observed Methylation", type = c("p", "smooth"), 
                        ylim=c(0, 1), cex.title=1.25, cex.axis=1.25, lwd = 3, degree = 1, span = 0.45, 
                        col = gg_color_hue(2), cex= rep(1.5, 1.5))

pdf(file.path(workingDir_carmona2017, "figs/Fig2A-DMR-24.pdf"), width = 8, height = 5)
plotTracks(list(dTrack_all, gtrack),
           groups = rep(c("450k", "RRBS"), each = 10), 
           main = "Chromosome 6",
           cex.legend = 1.25, cex.main = 1.1)
dev.off()
```


### Figure 2B

Compare beta values (CpGs) in regions methylated or not methylated 
using 450K platform and RRBS (Carmona et al. 2017 data). These 
are the same samples measured on two different platform technologies. 

```{r create-Figure2B}
#### 450k target data

data(offMethRegions)
data(onMethRegions)

dat.d0.target450k = methylCC:::.pick_target_positions(target_granges = gr_target450k, 
                                                      dmp_regions = offMethRegions)
mu0.target450k <- rowMeans(as.data.frame(mcols(dat.d0.target450k$dmp_granges)))

dat.d1.target450k = methylCC:::.pick_target_positions(target_granges = gr_target450k, 
                                                      dmp_regions = onMethRegions)
mu1.target450k <- rowMeans(as.data.frame(mcols(dat.d1.target450k$dmp_granges)))

#### RRBS target data
dat.d0.targetbs = methylCC:::.pick_target_positions(target_granges = gr_targetbs, target_object = M_targetbs, 
                                      target_cvg = cvg_targetbs, dmp_regions = offMethRegions)
mu0.targetbs <- rowMeans(as.data.frame(mcols(dat.d0.targetbs$dmp_granges)), na.rm = TRUE)
mu0.targetbs <- mu0.targetbs[mu0.targetbs != 0]

dat.d1.targetbs = methylCC:::.pick_target_positions(target_granges = gr_targetbs, target_object = M_targetbs, 
                                      target_cvg = cvg_targetbs, dmp_regions = onMethRegions)
mu1.targetbs <- rowMeans(as.data.frame(mcols(dat.d1.targetbs$dmp_granges)), na.rm = TRUE)
mu1.targetbs <- mu1.targetbs[mu1.targetbs != 1]

dat = data.frame("means" = c(mu0.target450k, mu0.targetbs, mu1.target450k, mu1.targetbs), 
                 "Platform" = as.factor(c(rep.int("450k", length(mu0.target450k)),
                                          rep.int("RRBS", length(mu0.targetbs)), 
                                          rep.int("450k", length(mu1.target450k)), 
                                          rep.int("RRBS", length(mu1.targetbs)))), 
                 "Regions" = factor(c(rep("Not methylated", length(c(mu0.target450k, mu0.targetbs))),
                                       rep("Methylated", length(c(mu1.target450k, mu1.targetbs)))), 
                                levels = c("Not methylated", "Methylated")))

p1 <- dat %>% ggplot(aes(x = means, colour = Platform)) + 
    stat_density(aes(colour = Platform, linetype = Regions), geom = "line",position="identity", size = 1.2) + 
    xlab("Methylation") + ylim(c(0, 130)) + xlim(c(0,1)) + 
theme(axis.text.x = element_text(size = rel(1.2)), axis.text.y = element_text(size = rel(1.2)), 
      title = element_text(size = rel(1.3)), panel.background = element_blank(), 
      legend.text = element_text(size = rel(1.2)), 
      legend.position = c(.95, .95),
      legend.justification = c("right", "top"),
      legend.box.just = "right",
      legend.margin = margin(6, 6, 6, 6)) 

pdf(file.path(workingDir_carmona2017,
              "figs/Fig2B-platformDependent-density.pdf"), width = 7, height = 5)
print(p1)
dev.off()
```


# Figure 3

Example region with a cell type-spcific Houseman CpG in 450K sample, 
but CpG is not observed in RRBS sample. Using cell type-specific regions
is a solution to this.Uses Carmona et al. (2017) data loaded from Figure 1

```{r create-Figure3}
library(Gviz)
gtrack <- GenomeAxisTrack()

dat_all <- merge(gr_target450k[,4], gr_targetbs[,1], all=TRUE)
genome(dat_all) <- "hg19"

# Combine everything and create individual tracks
expandbump <- 1e2

output <- find_dmrs(gr_target = NULL,
                    include_cpgs = FALSE, include_dmrs = TRUE)

kk <- subsetByOverlaps(output$regions_all, 
                       output$regions_houseman) + expandbump
ind = 11
tmp_flowSort = methylCC:::.pick_target_positions(gr_train450k, dmp_regions = kk[ind,])
dTrack_flowSort <- DataTrack(tmp_flowSort$probeGRanges, name = "Observed Methylation", type = c("p", "smooth"), 
                        ylim=c(0, 1), cex.title=1.25, cex.axis=1.25, lwd = 3, degree = 1, span = 0.75,
                        col = gg_color_hue(6), 
                        groups = pd0$CellType)

plotTracks(dTrack_flowSort)

tmp = methylCC:::.pick_target_positions(dat_all, dmp_regions = kk[ind,])
atrack_CT <- AnnotationTrack(subsetByOverlaps(output$regions_houseman, kk[ind,]), 
                             name = "CpG", stacking = "dense",
                             cex.title=1.1, cex.axis=1.1, rotation.title=0)

atrack_region <- AnnotationTrack(subsetByOverlaps(output$regions_all, kk[ind,]), 
                                 name = "DMR", stacking = "dense",
                                 cex.title=1.1, cex.axis=1.1, rotation.title=0)

ll <- methylCC:::.pick_target_positions(dat_all, dmp_regions = kk[ind,]- expandbump)
dTrack_all <- DataTrack(ll$probeGRanges, 
                         name = "Observed Methylation", type = c("p"), 
                        ylim=c(0, 1), cex.title=1.25, cex.axis=1.25, lwd = 3,
                        col = gg_color_hue(2), 
                        baseline = c(as.numeric(as.data.frame(mcols(ll$dmp_granges)))),
                        col.baseline = gg_color_hue(2), lty.baseline = 2, 
                        lwd.baseline = 3,
                        groups = rep(c("450k", "RRBS"), each = 1))

plotTracks(dTrack_all)

pdf(file.path(workingDir_carmona2017, 
              "figs/Fig3A-celltypespecificRegion.pdf"), width = 8, height = 5)
plotTracks(list(dTrack_flowSort, atrack_CT, atrack_region, gtrack),
           main = "Chromosome 14", 
           from = start(kk[ind,]), to = end(kk[ind,]),
           cex= 1.1, cex.legend = 1.25, cex.main = 1.1)
dev.off()


pdf(file.path(workingDir_carmona2017, 
              "figs/Fig3B-celltypespecificRegion.pdf"), width = 8, height = 5)
plotTracks(list(dTrack_all, atrack_region, gtrack),
           main = "Chromosome 14", 
           from = start(kk[ind,]), to = end(kk[ind,]),
           cex= 1.1, cex.legend = 1.25, cex.main = 1.1)
dev.off()
```


# Figure S1

```{r create-Figure4}
output <- find_dmrs(gr_target = NULL,
                    include_cpgs = FALSE, include_dmrs = TRUE)
dat <- as.data.frame(output$zmat)
dat <- dat[,c(4,2,3,1,5,6)]

pdf(file.path(workingDir_carmona2017, "figs/Figure4.pdf"), width = 5, height = 6)
rafalib::imagemat(dat[with(dat, order(-Bcell, -CD4T, -CD8T, -Gran, -Mono, -NK)), ], 
         main = "Cell type-specific DNAm profiles\nMethylated (black), Unmethylated (white)", 
         ylab = "Regions", xlab = " ", xaxt = "n", 
         cex.axis = 1.3, cex.lab = 1.3, cex.main = 1.3)
axis(1, at=1:6,labels=colnames(dat), srt=45, las = 2,
     cex.axis = 1.3, cex.lab = 1.3, cex.main = 1.3)
dev.off()
```




